### ISSUE TITLE
Cannot pass union of dict to function with Mapping input and generic key

### ISSUE THREAD

**Issue by Holt59** [microsoft/pyright#6005]
**Disclaimer:** Not sure if this is a "bug" or by design, submitting as a bug since this looks like one to me (or at least a missing feature).

**Describe the bug**

When calling a generic function an instance of a union of types, an error is thrown by `pyright`.

**Code or Screenshots**
```python
from __future__ import annotations
from typing import Mapping, TypeVar

_Key = TypeVar("_Key")

def f(x: Mapping[_Key, str]) -> Mapping[_Key, float]:
    ...

def g1(x: dict[int, str] | dict[tuple[int, float], str]) -> float:
    return sum(f(x).values())
```

Full error:

>  .\test.py
  .test.py:13:18 - error: Argument of type "dict[int, str] | dict[tuple[int, float], str]" cannot be assigned to parameter "x" of type "Mapping[_Key@f, str]" in function "f"
    Type "dict[int, str] | dict[tuple[int, float], str]" cannot be assigned to type "Mapping[int | tuple[int, float], str]"
      "dict[int, str]" is incompatible with "Mapping[int | tuple[int, float], str]"
        TypeVar "_KT@Mapping" is invariant
          Type "int" cannot be assigned to type "int | tuple[int, float]"
            "int" is incompatible with "tuple[int, float]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 

In my opinion, the type of `f(x)` can be deduced to `dict[int, float] | dict[tuple[int, float], float]`, so the function should be valid.

**VS Code extension or command-line**

Both:

- pylance: v2023.9.20
- pyright: 1.1.301

**Comment by Holt59** [comment]
@erictraut Thanks for the answer, is there a way to type `f` and `g1` so that no error is generated?

**Comment by erictraut** [comment]
I don't see a straightforward solution. Here's one that is a bit of a hack. It relies on a feature that both pyright and mypy implement. It's called "union expansion", and it's used only during overload matching. It's pretty expensive, which is why no Python type checker implements it for normal (non-overloaded) calls.

When performing overload matching, if no overload match can be found, arguments with union types are expanded into their individual subtypes. Overload matching is then performed again using these separate subtypes.

When you said that you thought this might be a missing feature, this is probably what you had in mind.

To use this in your code, you'd need to do something like this:

```python
from __future__ import annotations
from typing import Mapping, Never, TypeVar, overload

_Key = TypeVar("_Key")

@overload
def f(x: Mapping[_Key, str]) -> Mapping[_Key, float]:
    ...

@overload
def f(x: Never) -> Never:
    ...

def f(x: Mapping[_Key, str]) -> Mapping[_Key, float]:
    ...

def g1(x: Mapping[int, str] | Mapping[tuple[int, float], str]) -> float:
    return sum(f(x).values())
```

It's not especially pretty, but it gets the job done — at least with mypy and pyright.

Note that union expansion is not part of the Python typing spec, and I don't think other static or runtime type checkers implement it.
